core/src/main/java/org/infinispan/container/offheap/OffHeapDataContainer.java

```diff
@@ -154,7 +155,7 @@ protected void checkDeallocation() {
       while (address != 0) {
          long nextAddress = offHeapEntryFactory.getNext(address);
          InternalCacheEntry<WrappedBytes, WrappedBytes> ice = offHeapEntryFactory.fromMemory(address);
-         if (wrappedKey.equalsWrappedBytes(ice.getKey())) {
+         if (wrappedKey.equalsWrappedBytes(ice.getKey()) && !ice.isExpired(timeService.wallClockTime())) {
```

Comment:
There are a few other places this is need. I would advise renaming `entryStream` to `entryStreamIncludingExpired` and make a new `entryStream` that invokes `entryStreamIncludingExpired` that filters out expired entries. Then we can call the proper method. Currently only the `iteratorIncludingExpired` should have visibility to expired entries.

We also have to add an `isExpired` check in the `executeTask`, `containsKey` and `remove` methods as they all currently operate upon expired entries. For the `containsKey` we should add a method to the `OffHeapEntryFactory` class that just returns if the key is expired or not. Then it only needs to read the type to figure out what expiration variables to read instead of the entire object.

If you want I can handle this, but I figure you might enjoy doing this :)